tiledb/sm/subarray/range_subset.h

```diff
@@ -0,0 +1,328 @@
+  //  */
+  // // virtual Status add_range(
+  // //   std::vector<std::vector<Range>>& ranges, Range&& new_range) = 0;
+
+  /**
+   * Adds a range to the range manager without performing any checkes. If a
+   * default strategy is set, then first update the range strategy.
+   *
+   * @param ranges The current ranges in the subarray (remove after
+   refactor).
+   * @param new_range The range to add.
+   */
+  virtual Status add_range_unsafe(const Range& range) = 0;
+
+  virtual const Range& get_range(const uint64_t range_index) const = 0;
+
+  virtual const std::vector<Range>& get_ranges() const = 0;
+
+  /**
+   * Returns ``true`` if the current range is the default range.
+   *
+   * The default range is the full range for the dimension the range is
+   * managing.
+   **/
+  virtual bool is_default() const = 0;
+
+  /** Returns ``true`` if the range subset is the empty set. */
+  virtual bool is_empty() const = 0;
+
+  /**
+   * Returns ``false`` if the subset contains a range other than the default
+   * range.
+   **/
+  virtual bool is_set() const = 0;
+
+  /**
+   * Returns ``true`` if there is exactly one unary ranage in the subset.
+   */
+  virtual bool is_unary() const = 0;
+
+  /** Returns the number of distinct ranges stored in the range manager. */
+  virtual uint64_t num_ranges() const = 0;
+
+  /**
+   * Sorts the ranges in the range manager.
+   *
+   * @param compute_tp
+   */
+  virtual Status sort_ranges(ThreadPool* const compute_tp) = 0;
+};
+
+template <typename T, Datatype D, bool CoalesceAdds>
+class RangeSubset : public RangeSubsetBase {
+ private:
+  using AddStrategy = detail::AddStrategy<CoalesceAdds, T>;
+  using SortStrategy = detail::SortStrategy<D, T>;
+
+  /** Maximum possible range. */
+  Range full_range_;
+
+  /**
+   * If ``true``, the range contains the full domain for the dimension (the
+   * default value for a subarray before any other values is set. Otherwise,
+   * some values has been explicitly set to the range.
+   */
+  bool is_default_ = true;
+
+  /** Stored ranges. */
+  std::vector<Range> ranges_;
+
+ public:
+  /** Disable default constructor. */
+  RangeSubset() = delete;
+
+  /**
+   * Constructor for the RangeSubset.
+   *
+   * @param full_range The range this is a subset of.
+   * @param is_default Flag indicating if this is a default subset. The default
+   * subset is the full range.
+   */
+  RangeSubset(const Range& full_range, bool is_default)
+      : full_range_(full_range)
+      , is_default_(is_default)
+      , ranges_() {
+    if (is_default)
+      ranges_.emplace_back(full_range);
+  };
+
+  /** Destructor. */
+  ~RangeSubset() = default;
+
+  //  Status add_range(
+  //      Range&& range,
+  //      bool error_on_oob) {
+  //    if (!error_on_oob)
+  //      RETURN_NOT_OK(adjust_range_oob(&new_range));
+  //    RETURN_NOT_OK(check_range(&new_range));
+  //    add_range_unsafe(ranges, new_range);
+  //  };
+
+  Status add_range_unsafe(const Range& range) override {
```

Comment:
What's unsafe about this?

Reply:
No check that the range is in fact a subset of the full range.

Reply:
That's not unsafe, it's merely unrestricted.

Let S be a set that's the internal state of this class. Let A be the argument set, and let B be the domain set.
There are two operations possible:
* S := S \setunion A
* S := S \setunion (A \setintersect B)
The first operation is what this function is doing. The second would preserve an invariant that S \subset B. In order to call an operation on this class "unsafe", it would need B as part of its object state. So, from the point of view of `RangeSubset`, it's just an ordinary incremental union operation.

There might be good reason to have another class `RestrictedRangeSubset` whose constructor takes a set B as an argument. This class would have `add_range` implement the second function. That class could have a `set_range_unsafe` that would need to set an internal flag that the subset invariant might not be valid.

The point is that from a class that doesn't have a restriction set, it's not an unsafe operation insofar as this class is concerned. The user should use it with such in mind, but it shouldn't have a name that reflects one way of using it.

Reply:
I think you are confusing "set" and "subset". You cannot have a subset without some notion of what the full set is. Moreover, this method is unsafe. Consider a subset S of a set B. If A is not a subset of B, then (A union S) if not a subset of B. This contradicts the implicit "is subset" definition of the class.

Using your notation above, this class does have B as an argument. It's called `full_range_`. Would it help clarify things if I renamed `full_range_`? If so, do you have a preferred name in mind?

Reply:
The full set I was referring to is the type set, which here is all possible ranges. I admit that this might have been confusing. On the other hand. it wasn't clear at all that `full_range_` was anything other than an implementation detail involving defaults.

> preferred name

The class name is where we need to start for this. There are a few consistent possibilities. All have these properties:
- The class has as stored member set S on which it exposes set operations.
- The class has a reference to a set B fixed at construction.

Different possibilities for the class:
1. A class that would just hold S and B in the same place for convenience, but would not do anything to enforce any particular relationship between them. In particular, this kind of class would require the user to keep track of any relationship between S and B that they care about. When you retrieve S, nothing at all about its relationship to B is known.
2. A class with an invariant that S \subset B. This would require an intersection operation during `add_range`. When you retrieve S, S will always be a subset of B.
3. A class with a semivariant that S \subset B, meaning there's an exposed flag that marks when S \subset B is true. This class would have an invariant "is_S_subset_of_B \implies S \subset B`. If "is_S_subset_of_B" is false, you can't conclude anything.

Item 1 describes the current class. Item 2 could readily be called `RangeSubset`, since it's enforcing an invariant that names the class. Item 3 should probably be called `RangePotentialSubset` since it's only got a semivariant name. In the current state of the class, I'd say a good name would be `RangeSetAndSuperset`.

It's too much work in this PR to implement something that deserves the name `RangeSubset`. Since it's not clear at this stage where responsibility will lie for ensure the subset predicate, it would prudent to plan on this class undergoing a subsequent name change alongside any change in responsibilities. I don't think there's any problem with giving a clunky name to clunky just-being-refactored code and indeed I see an advantage to it, because you're not claiming anything for it that's not there.

Now, having said all this, we will probably want something like `RangeSet` that has no superset concept anywhere in it. No need to add that in the PR for expediency. One non-obvious difficulty of separating them out will mean that you need to split `AddPolicy` into two pieces: one for the set itself and one to deal with the subset restriction. So don't do all that in this PR.

For all this talk, the main changes to the PR are (1) naming and (2) documentation. I'd rename the class along the lines above. While it would be pretty easy to turn it into a `RangePotentialSubset`, I think your predilection to not make significant behavior changes in this PR is fine. `full_range_` should probably be `superset_`. `add_range_unsafe` isn't unsafe if the class isn't making any guarantees about a relationship between S and B, so it should just be `add_range`.

As for documentation, it should be explicit about a few things:
* The maturity state of the class, in particular, that it's transitional toward a rewriting goal.
* The lack of any actual relationship between S and B. There's a relationship goal, of course, but it's not present yet.
* A short statement about what the aspirational goals are, namely, that it no longer be a mere-convenience class.

Reply:
I will update the documentation. This class is currently being used as though it satisfies Item 2, and so that is where I want to get it before this refactor is finished. Given that, do you think it would be unreasonable to leave the file name as RangeSubset.* and just change the name of the class itself?